staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go

```diff
@@ -0,0 +1,127 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package filters
+
+import (
+	"fmt"
+	"net/http"
+	"runtime"
+
+	// TODO: decide whether to also generate the old metrics, which
+	// categorize according to mutating vs readonly.
+
+	// "k8s.io/apiserver/pkg/endpoints/metrics"
+
+	fcv1a1 "k8s.io/api/flowcontrol/v1alpha1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	apirequest "k8s.io/apiserver/pkg/endpoints/request"
+	utilflowcontrol "k8s.io/apiserver/pkg/util/flowcontrol"
+	"k8s.io/klog"
+)
+
+// WithPriorityAndFairness limits the number of in-flight
+// requests in a fine-grained way.
+func WithPriorityAndFairness(
+	handler http.Handler,
+	longRunningRequestCheck apirequest.LongRunningRequestCheck,
+	fcIfc utilflowcontrol.Interface,
+) http.Handler {
+	if fcIfc == nil {
+		klog.Warningf("priority and fairness support not found, skipping")
+		return handler
+	}
+
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		requestInfo, ok := apirequest.RequestInfoFrom(ctx)
+		if !ok {
+			handleError(w, r, fmt.Errorf("no RequestInfo found in context"))
+			return
+		}
+		user, ok := apirequest.UserFrom(ctx)
+		if !ok {
+			handleError(w, r, fmt.Errorf("no User found in context"))
+			return
+		}
+
+		// Skip tracking long running requests.
+		if longRunningRequestCheck != nil && longRunningRequestCheck(r, requestInfo) {
+			klog.V(6).Infof("Serving RequestInfo=%#+v, user.Info=%#+v as longrunning\n", requestInfo, user)
+			handler.ServeHTTP(w, r)
+			return
+		}
+
+		note := func(fs *fcv1a1.FlowSchema, pl *fcv1a1.PriorityLevelConfiguration) {
+		}
+		var served bool
+		execute := func() {
+			served = true
+			// Serve the request, but asynchronously, and return from here
+			// as soon as either the request is finished or the context is
+			// canceled.  The logic here is heavily cribbed from the
+			// timeout filter.
+			timedOut := ctx.Done()
+
+			// gets sent to exactly once, with the result of recover()
+			// for serving the request
+			resultCh := make(chan interface{})
+
+			go func() { // serve the request, with recovery from panics
+				defer func() {
+					err := recover()
+					if err != nil && err != http.ErrAbortHandler {
+						// Same as stdlib http server code. Manually allocate stack
+						// trace buffer size to prevent excessively large logs
+						const size = 64 << 10
+						buf := make([]byte, size)
+						buf = buf[:runtime.Stack(buf, false)]
+						err = fmt.Sprintf("%v\n%s", err, buf)
+					}
+					resultCh <- err
+				}()
+				handler.ServeHTTP(w, r)
+			}()
+
+			select {
+			case err := <-resultCh:
+				if err != nil {
+					panic(err)
+				}
+			case <-timedOut:
```

Comment:
I talked this over with @lavalamp this afternoon.

While it's tempting to free up a concurrency share immediately when a request times out, we actually think that should not happen: the concurrency limit should be a limit on how much _actual work_ the apiserver is doing, not how much _useful_ work it's doing. (This corresponds with the present-day behavior of the MaxInFlight handler, which also does not release the in-flight request token until the downstream handler has actually finished, whether its work will be fruitful or not.)

We therefore don't need to fork the above goroutine at all, AFAICT; execute() should just call handler.ServeHTTP in the foreground. The existing timeout handler will take care of returning a 504, and we don't want the queueset to start new work until we're well and truly done consuming apiserver resources for this one.

Reply:
I will not claim to know this about kube apiservers, but my experience with other servers is that requests taking a really long time are almost always stuck doing nothing.

Reply:
> the concurrency limit should be a limit on how much actual work the apiserver is doing, not how much useful work it's doing.

+1. consider users flooding apiserver w/ a bunch of heavy/unpaginated LIST calls, those on-going/hanging LIST calls running background should always keep their concurrencies until the work is actually done. while OTOH, if everything runs foreground, imagine an unresponsive third-party AA servers in the cluster can potentially run out the quota.

Reply:
> I will not claim to know this about kube apiservers, but my experience with other servers is that requests taking a really long time are almost always stuck doing nothing.

Yeah, that's definitely a common pattern but unfortunately doesn't match my experience with stuck apiservers: there's a fair amount of real work (including, occasionally, waiting on self-client requests) that proceeds very slowly in overload situations. Ideally most things would be much better at quickly stopping what they're doing when the context is cancelled and that is a separate crusade we should go on some day.

Reply:
I have removed the code for reacting to cancelation during execution.